File: server/schema.py
import random
import time
import uuid
from utils import get_semantic_location

class User:
    connections = []
    def __init__(self, name:str=None, email:str=None, image:str=None, setup_dict=None):
        if(setup_dict is not None):
            #careful!
            for key, value in setup_dict.items():
                setattr(self, key, value)
            return
        self.id = str(uuid.uuid4())
        self.name = name
        self.email = email
        self.image = image
        self.createdAt = time.time()
        self.lastPolled = self.createdAt
        self.connections = []
        self.paired_with = []
        self.connection_today = -1
        self.last_paired = self.createdAt

class Message:
    def __init__(self, userid:str, content:str, timeEdited:float, setup_dict=None):
        if setup_dict is not None:
            # careful!
            for key, value in setup_dict.items():
                setattr(self, key, value)
            return
        self.id = uuid.uuid4()
        self.user_id = userid
        self.content = content
        self.timeCreated = time.time()
        self.timeEdited = timeEdited

class Chat:
    messages:list[Message] = []
    def add_message(self, userid:str, content:str):
        self.messages.append(Message(userid, content))

class Connection:
    should_continue = (False, False)
    has_begun = False
    def __init__(self, user1: User, user2: User, geo_location:tuple[float, float], setup_dict=None):
        if setup_dict is not None:
            # careful!
            for key, value in setup_dict.items():
                setattr(self, key, value)
            return
        self.id = uuid.uuid4()
        self.created = time.time()
        self.expiry = time.time() + 30 * 24 * 3600
        #location is a lat + long coordinate pair. But is this right? may want to change to a semantic location ie "Red Square"
        self.geo_location = geo_location
        self.full_location = get_semantic_location(geo_location)
        self.semantic_location = get_semantic_location(geo_location).get("display_name")
        self.user_ids = (user1.id, user2.id)
        #todo: is this right?
        self.chat = Chat()
        user1.connections.append(self.id)
        user2.connections.append(self.id)

    def check_for_expiry(self) -> bool:
        return time.time() > self.expiry and self.should_continue.__contains__(False)

File: composeApp/src/commonMain/kotlin/compose/project/click/click/schema.kt
package compose.project.click.click

import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable

/**
 * Note: These data classes are Kotlin translations of the provided Python schema.
 * Business logic and default value generation (like UUIDs or timestamps) from the original
 * Python classes have been separated. You would typically handle ID generation and
 * timestamping in your app's business logic layer (e.g., a repository or view model).
 */

@Serializable
data class GeoLocation(
    val lat: Double,
    val lon: Double
)

@Serializable
data class LoginObject(
    val jwt: String,
    val refresh: String,
    val user: User
)

/**
 * Represents a user in the system.
 * Translated from the Python `User` class.
 */
@Serializable
data class User(
    val id: String,
    val name: String? = null,
    val email: String? = null,
    val image: String? = null,
    val createdAt: Instant? = null,
    val lastPolled: Instant? = null,
    val connections: List<String> = emptyList(),
    val paired_with: List<String> = emptyList(),
    val connection_today: Int = -1,
    val last_paired: Instant? = null
)

/**
 * Represents a single message in a chat.
 * Translated from the Python `Message` class.
 */
@Serializable
data class Message(
    val id: String,
    val user_id: String,
    val content: String,
    val timeCreated: Instant? = null,
    val timeEdited: Instant? = null
)

/**
 * Represents a chat session containing messages.
 * Translated from the Python `Chat` class.
 */
@Serializable
data class Chat(
    val messages: List<Message> = emptyList()
)

/**
 * Represents a connection between two users.
 * The Python `get_semantic_location` is not translated; you'll need a Kotlin implementation for that.
 */
@Serializable
data class Connection(
    val id: String,
    val created: Instant? = null,
    val expiry: Instant? = null,
    val geo_location: GeoLocation,
    // Represents the full JSON object from a semantic lookup, equivalent to `get_semantic_location()`
    val full_location: Map<String, String>? = null,
    // Represents the 'display_name' from the semantic location lookup
    val semantic_location: String? = null,
    val user_ids: List<String>,
    val chat: Chat = Chat(),
    val should_continue: List<Boolean> = listOf(false, false),
    val has_begun: Boolean = false
)